### PR TITLE
[ci] Port software build job from Azure to GitHub

### PR DESCRIPTION
**Pull request by jwnrt**
Moves both the `sw_build` and `sw_test` jobs from Azure into one `sw_build_test` job in GitHub Actions.

There's a small improvement in execution time here for a cold build. I'll check the time when the results are cached once this merges.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -241,3 +241,72 @@ jobs:
...
+
+  sw_build_test:
+    name: Build and test software
+    runs-on: ubuntu-22.04-vivado
+    timeout-minutes: 120
+    needs: quick_lint
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0 # Required so we can lint commit messages.
```
Comment (nbdd0121) [inline .github/workflows/ci.yml:253]: Do we still need this for sw build & test?

Comment (jwnrt) [reply .github/workflows/ci.yml:253]: The comment is wrong, but we do need this for the bitstream cache to work as well. I'm going to remove this and see if this job does any splicing.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -241,3 +241,72 @@ jobs:
...
+            --test_output=errors                                      \
+            --define DISABLE_VERILATOR_BUILD=true                     \
+            --test_tag_filters=-broken,-cw310,-verilator,-dv,-silicon \
+            --target_pattern_file="$target_pattern_file"
+      - name: Publish test results
+        uses: actions/upload-artifact@v4
+        with:
+          name: test_results
+          path: "**/test.xml"
+          overwrite: true
```
Comment (nbdd0121) [inline .github/workflows/ci.yml:309]: This is what I currently have for FPGA jobs:

```yaml
- name: Publish Bazel test results
  uses: mikepenz/action-junit-report@ec3a351c13e080dc4fa94c49ab7ad5bf778a9668 # v5
  if: ${{ !cancelled() }}
  with:
    report_paths: 'bazel-out/**/test.xml'
```

Perhaps we can have a composite action that merge junit, upload the xml and publish junit report?

Comment (jwnrt) [reply .github/workflows/ci.yml:309]: Yeah, that would be useful

**Review by nbdd0121 — APPROVED** [review]
Failing test seems unrelated

**Pull request merged**